jetty-server/src/main/java/org/eclipse/jetty/server/HttpInput.java

```diff
@@ -190,633 +89,348 @@ public Interceptor getInterceptor()
-                        if (_channelState.isResponseCommitted())
-                            _channelState.getHttpChannel().abort(bad);
-                        throw bad;
-                    }
-                }
-            }
-
-            // Consume content looking for bytes to read
-            while (true)
-            {
-                Content item = nextContent();
-                if (item != null)
-                {
-                    l = get(item, b, off, len);
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("{} read {} from {}", this, l, item);
-
-                    // Consume any following poison pills
-                    if (item.isEmpty())
-                        nextInterceptedContent();
-                    break;
-                }
-
-                // No content, so should we block?
-                if (!_state.blockForContent(this))
-                {
-                    // Not blocking, so what should we return?
-                    l = _state.noContent();
-
-                    if (l < 0)
-                        // If EOF do we need to wake for allDataRead callback?
-                        wake = _channelState.onReadEof();
-                    break;
-                }
-            }
-        }
-
-        if (wake)
-            wake();
-        return l;
+            _contentProducer.setInterceptor(new ChainedInterceptor(currentInterceptor, interceptor));
     }
 
     /**
-     * Called when derived implementations should attempt to produce more Content and add it via {@link #addContent(Content)}. For protocols that are constantly
-     * producing (eg HTTP2) this can be left as a noop;
-     *
-     * @throws IOException if unable to produce content
+     * Called by channel when asynchronous IO needs to produce more content
      */
-    protected void produceContent() throws IOException
+    public void asyncReadProduce()
     {
+        if (LOG.isDebugEnabled())
+            LOG.debug("asyncReadProduce {}", _contentProducer);
+        produceRawContent();
     }
 
     /**
-     * Called by channel when asynchronous IO needs to produce more content
+     * Adds some content to this input stream.
      *
-     * @throws IOException if unable to produce content
+     * @param content the content to add
      */
-    public void asyncReadProduce() throws IOException
+    public void addContent(Content content)
     {
-        synchronized (_inputQ)
+        if (LOG.isDebugEnabled())
+            LOG.debug("addContent {} {}", content, _contentProducer);
+        if (_firstByteTimeStamp == Long.MIN_VALUE)
         {
-            produceContent();
+            _firstByteTimeStamp = System.nanoTime();
+            if (_firstByteTimeStamp == Long.MIN_VALUE)
+                _firstByteTimeStamp++;
         }
+        _contentProducer.addContent(content);
+        if (isAsync())
+            _channelState.onContentAdded();
```

Comment:
I think the problem here is that we may wake up if we receive raw content that is fully consumed by an interceptor.
I think that either:
 - addContent should attempt to do a transform and return a boolean to indicate if there is some; OR
 - in the isAsyncCase you try calling nextNonEmptyContent to check if there is a non zero return.

Reply:
Or this?
```
if (isAsync() && _contentProducer.available(this::produceContent) > 0)
    _channelState.onContentAdded();
```

Reply:
That's a nice way of doing it!   
Don't worry about the L1 cache, because the thread that does the transformation is doing so from with a HttpChannel.handle, which will see the changed state on unhandle and loop around to call the onDataAvailable callback, which will then call read and find all that nice tranformed data in its L1 cache!